src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Automation/AbstractExtenderCATIDProvider.cs

```diff
@@ -0,0 +1,112 @@
+            {
+                case ExtenderCATIDType.HierarchyExtensionObject:                                // IVsHierarchy.GetProperty(VSITEMID.Root, VSHPROPID_ExtObjectCATID)
+                case ExtenderCATIDType.AutomationProject:                                       // DTE.Project
+                    return GetExtenderCATID(ExtendeeObject.Project);
+
+                case ExtenderCATIDType.HierarchyBrowseObject:                                   // IVsHierarchy.GetProperty(VSHPROPID_BrowseObjectCATID)
+                case ExtenderCATIDType.ProjectBrowseObject:                                     // EnvDTE.Project.Properties
+                    return GetExtenderCATID(ExtendeeObject.ProjectBrowseObject);
+
+                case ExtenderCATIDType.ConfigurationBrowseObject:                               // IVsCfgProvider2.GetCfgProviderProperty(VSCFGPROPID_IntrinsicExtenderCATID)/DTE.Configuraton
+                    return GetExtenderCATID(ExtendeeObject.Configuration);
+
+                case ExtenderCATIDType.HierarchyConfigurationBrowseObject:                      // IVsHierarchy.GetProperty(VSHPROPID_CfgBrowseObjectCATID)
+                case ExtenderCATIDType.ProjectConfigurationBrowseObject:                        // EnvDTE.Configuration.Properties
+                    return GetExtenderCATID(ExtendeeObject.ConfigurationBrowseObject);
+                 
+                case ExtenderCATIDType.AutomationProjectItem:                                   // EnvDTE.ProjectItem
+                    return GetExtenderCATID(ExtendeeObject.ProjectItem);
+
+                case ExtenderCATIDType.AutomationReference:                                     // VSLangProject.Reference
+                case ExtenderCATIDType.ReferenceBrowseObject:                                   // EnvDTE.ProjectItem.Properties (when reference)
+                    return GetExtenderCATID(ExtendeeObject.ReferenceBrowseObject);
+
+                case ExtenderCATIDType.FileBrowseObject:                                        // EnvDTE.ProjectItem.Properties (when file)
+                    return GetExtenderCATID(ExtendeeObject.FileBrowseObject);
+
+                case ExtenderCATIDType.AutomationFolderProperties:                              // FolderProperties
+                case ExtenderCATIDType.FolderBrowseObject:                                      // EnvDTE.ProjectItem.Properties (when folder)
+                    return GetExtenderCATID(ExtendeeObject.FolderBrowseObject);
+
+                default:
+                case ExtenderCATIDType.Unknown:                                                 // EnvDTE.ProjectItem.Properties (when not file, folder, reference)
+                    BCLDebug.Assert(extenderCATIDType == ExtenderCATIDType.Unknown, $"Unrecognized CATID type {extenderCATIDType}");
+                    return null;
+            }
+        }
+
+        protected abstract string GetExtenderCATID(ExtendeeObject extendee);
```

Comment:
Should this take an `IProjectTree` for the node for when the second issue above is fixed?

Reply:
No, 'IProjectTree' is not a good representation of all the objects we're referring to, some, like a configuration, don't have a representation in the tree `at all`. The DTE objects themselves are actually a better representation at this point.